### DISCUSSION TITLE
POST Json data to HTTP Listener

### DISCUSSION THREAD

**Discussion by mankowitz** [nextgenhealthcare/connect#5142] [General]
I am trying to post data to an HTTP Listener, but whenever I set the data source to JSON, I get the following error (see below)

I have attached the sample channel, but the key items are these:
Source Connector Data Types: JSON/JSON
Connector Type: Http Listener
HTTP Listener Settings:  Message Content: Plain body; 
Destination is just a file writer

I'm using Mirth Connect (open source) 4.0.0

```
Transformer error
ERROR MESSAGE: Error evaluating transformer
com.mirth.connect.server.MirthJavascriptTransformerException: 
CHANNEL:	http listener test
CONNECTOR:	Destination 1
SCRIPT SOURCE:	
SOURCE CODE:	
94:         }
95:         eval('importPackage(' + Packages.java.lang.Class.forName(className).getPackage().getName() + ')');
96:     }
97: }
98: function doScript() {
99: msg = JSON.parse(connectorMessage.getTransformedData());
100: function doFilter() { phase[0] = 'filter'; return true; }function doTransform() { phase[0] = 'transformer'; logger = Packages.org.apache.log4j.Logger.getLogger(phase[0]);
101: 
102: 
103: if ('xml' === typeof msg) {
LINE NUMBER:	99
DETAILS:	SyntaxError: Empty JSON string
	at 41785865-1e6b-42bb-86a7-b6e64eb12ed3_JavaScript_Filter_Transformer_1:99 (doScript)
	at 41785865-1e6b-42bb-86a7-b6e64eb12ed3_JavaScript_Filter_Transformer_1:126
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:234)
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:186)
	at com.mirth.connect.server.util.javascript.JavaScriptTask.call(JavaScriptTask.java:113)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```
[http listener test.zip](https://github.com/nextgenhealthcare/connect/files/8501393/http.listener.test.zip)

The data payload looks like this (It's not PHI)
```json
{"patient":{"id":38,"first_name":"STACY","last_name":"BARCON","site_id":1190,"site_mrn":"SMD38","created_at":"2018-09-16T04:24:12.000000Z","updated_at":"2021-10-05T13:08:10.000000Z","deleted_at":null,"completed_at":"2018-08-05T04:00:00.000000Z","completed_by":52,"chief_complaint_id":null,"created_by":52,"assigned_to":52,"clinical_info":null,"pt_location":null,"pmd_id":null,"diagnosis":"foliculitis","interventions":"keflex and warm compresses","outcome_id":"obs","admitted":0,"transferred_48h":null,"vital_signs":"no fever","note_in_emr":1,"orders_signed":1,"pmd_notified":1,"video_used":0,"assigned_at":null,"callback_phone_number":"(202) 456-1414","priority":0,"chief_complaint_name":"","pmd_name":null,"scribe_id":null,"ap_id":null,"doctor_id":52,"created_year_month_idx":"2018-09-01","dob":"1976-03-30","is_temporary":0,"mpi_id":63,"urgency_id":null,"histories":null,"follow_up":null,"billable":0,"pt_info":null,"transfer_info":"{\"contacted\": 0}","length_of_stay":"","length_of_stay_level":1,"chief_complaint_text":"Unk","pmd_text":"Unk","bravado_status":{"pharmacyId":null,"encounterId":null,"patientId":null,"needs":[]},"mpi":{"id":63,"first_name":"STACY","last_name":"BARCON","created_at":"2018-06-12T09:13:53.000000Z","updated_at":"2021-11-14T03:32:58.000000Z","dob":"1976-03-30T05:00:00.000000Z","sex":"M","pt_info":"{\"consent\": {\"date\": \"2020-01-01\", \"method\": \"administrative\", \"obtained_by\": \"agency\"}}","pmd_id":null,"deleted_at":null,"deleted_id":null,"notifications":null,"histories":null},"chief_complaint":null,"pmd":null,"site":{"id":1190,"created_at":"2020-03-08T02:04:00.000000Z","updated_at":"2021-11-14T02:58:56.000000Z","deleted_at":null,"site_info":null,"zoom_license":1,"medical_director_id":null,"client_ehr_id":4,"phone_number":"631-744-7158-get extension","local_facilities":"[{\"name\": \"John T. Mather Memorial Hospital\", \"type\": \"ED\", \"phone\": \"631-473-1320\"}, {\"name\": \"Stony Brook Hospital\", \"type\": \"ED\", \"phone\": \"631-444-2499\"}, {\"name\": \"St. Catherine of Siena\", \"type\": \"ED\", \"phone\": \"631-862-3000\"}, {\"name\": \"St. Charles Hospital\", \"type\": \"ED\", \"phone\": \"(631) 474-6000\"}, {\"name\": \"Peconic Bay Medical Center\", \"type\": \"ED\", \"phone\": \"(631) 548-6000\"}]","eko_link":"https://app.ekodevices.com/livestream/fb8eee0cdeff7baebc7cf7b3","external_links_json":null,"short":null,"fax_number":null,"on_chart_complete":null,"chart_complete_req_fields":null,"agency_id":30,"state":"NY","county":"Suffolk","site_type_id":1,"hide_dynamic_zoom_link":0,"timezone":"America/New_York","primary_name":null,"secondary_name":"Shoreham","address1":"283 NY-25A","address2":null,"city":"Shoreham","zip":"11786","alias_info":"null","billable":null,"name":" - Shoreham: 283 NY-25A , Shoreham, NY 11786"}},"oid":"2.16.840.1.113883.3.9447.1.2"}
```

**Comment by pacmano1** [comment]
That JSON cuts and pastes oddly to some editors.  Like a special character after "consent".

Can you upload that JSON as file attachment or post it somewhere as a file?

Also you formatted your code/post with the channel zip link as part of the error, folks might not be seeing your download link.

**Reply by mankowitz** [reply]
The JSON rendering is another problem, not related to mirth. In `patient.mpi.pt_info`, the text there is an escaped json string. The rest of the json payload should be parsed

**Reply by pacmano1** [reply]
I just meant it looked like there was some unprintable character when I copied your post to sublime text editor.   I know what escaped JSON looks like.

![image](https://user-images.githubusercontent.com/44065187/164043725-51f9107a-45e6-434f-8723-502244183995.png)

**Comment by eduardogr1984** [comment]
### -[EDIT]: I found a solution to my case, can be seen in a post below, click the "View more" link if you don't get to see it.
**Solution comment: https://github.com/nextgenhealthcare/connect/discussions/5142#discussioncomment-2615301**

Just having the same problem with Mirth Connect v3.10.1. My source connector is a HTTP Listener with JSON/JSON DataTypes: I've even tried different combinations in the Source connector "Message content, Response Content Type, Response Data Type" parameters with no succcess. Tried to set different "Content-Type" in the request headers, making it match the type in the Source connector but still the same problem. Tried with transformers and without them too.

What is interesting is that if I send a Message (through the Mirth UI) into the channel with an example payload of {"asd": "123"} it is correctly recognized, but If I send exactly the same content into a POST request from outer system (Postman), the error shows all the time.
It is like the body content isn't been recognized correctly when it is sent from an outer request...

Please give me some help. If someone wants to see anythig about my configuration just tell me. Thanks in advance.

**Reply by pacmano1** [reply]
Post a minimal export of your channel that replicates the behavior. Also please describe the exact configuration of Mirth including whether or not it is in a container, whether or not there are load balancers in front of it and or web application firewalls / IDS / IPS that might interfere with the traffic.  if any of the latter things exist, please make sure you have replicated this issue in environment that has none of these potential systems that might interfere with the transaction.

**Reply by eduardogr1984** [reply]
Hi. **No container nor dockered, local installed with no balancers, no firewalls, etc**. 

My guess is that, somehow the **content is not been recognized when sending a request from an outer system** (Postman in this case), because **it works when the payload content is recognized by sending an internal message** to the channel, but I can't figure what is happening. 

I uploaded several configuration and results screenshots in case it helps. Thanks in advance.

![2022-04-21_18h30_04](https://user-images.githubusercontent.com/23081721/164512357-086b8f5b-14b9-446d-b93b-9c3ec8092458.jpg)

![2022-04-21_18h30_33](https://user-images.githubusercontent.com/23081721/164512382-85410e31-dd39-4d52-91dc-f5925f8f55ee.jpg)

![2022-04-21_18h30_55](https://user-images.githubusercontent.com/23081721/164512391-c180fd89-f940-4cdf-ad3b-79ab3c48633a.jpg)

![2022-04-21_18h36_39](https://user-images.githubusercontent.com/23081721/164512411-9a80df07-4438-4a51-8efc-efc0953550d9.jpg)

![2022-04-21_18h38_11](https://user-images.githubusercontent.com/23081721/164512421-6463ee37-b47b-45c9-91f8-1d39ae6ab157.jpg)

![2022-04-21_18h38_33](https://user-images.githubusercontent.com/23081721/164512442-82e3bd70-9a54-49b4-b1fc-6fb578beca17.jpg)

![2022-04-21_18h39_09](https://user-images.githubusercontent.com/23081721/164512458-4d8d0261-6058-4823-aa38-f2957c54e439.jpg)

![2022-04-21_18h39_31](https://user-images.githubusercontent.com/23081721/164512469-604c1572-fb56-415b-a471-c15c10e26bba.jpg)

![2022-04-21_18h39_47](https://user-images.githubusercontent.com/23081721/164512483-4c3c548d-c002-47a3-9bf6-b0b3ec513350.jpg)

**Reply by pacmano1** [reply]
Please post an export of the channel, I asked that already.   Also for a quick test, can you stop using ipv6 for all of it and use v4 with explicit IP addresses?

**Reply by eduardogr1984** [reply]
There you have the channel Export. I've tried disabling IPv6 protocol from my network adapter and specified my adapters IP, but all the same happened. Also tried changing the Postman request Content-Type to 'application/fhir+json' and many other configurations, but no changes at all.

[FHIR Test Channel.zip](https://github.com/nextgenhealthcare/connect/files/8537619/FHIR.Test.Channel.zip)

Thanks.

**Reply by eduardogr1984** [reply]
### Finally found a solution to my problem.

At least in my case, the problem is that I'm using a "Base context path" in my source HTTP Listener (also downloaded the originals poster channel and checked that he had it too), and the URL provided ends with a "/ ", which can be seen into the "HTTP URL" field, note that there is a trail slash at the end of the URL:
![2022-04-22_13h40_04](https://user-images.githubusercontent.com/23081721/164712991-190b4d2f-3698-4d2a-bd6a-54041ca8e540.jpg)

**It seems that the trail slash is important**, as can be seen in the following topics (search for the answers from "narupley"):
https://forums.mirthproject.io/forum/mirth-connect/support/11350-how-can-i-setup-a-http-listener-to-support-put-rest-call
https://forums.mirthproject.io/forum/mirth-connect/support/16577-http-listener

When a Base Context Path is used, when calling the endpoint the trail slash must be explicitly specified too. When I did it that way all worked fine:
![2022-04-22_13h44_23](https://user-images.githubusercontent.com/23081721/164713349-fb3fcd0a-4596-417b-9a06-6690d49c2280.jpg)

![2022-04-22_13h53_16](https://user-images.githubusercontent.com/23081721/164713370-0833d157-6d3e-4003-8bbb-6468176b24b1.jpg)

It is interesting that this is not happening when not using the "Base Context Path", although the slash is present in the HTTP Listener URL, it is not a problem to exclude it.

Anyway, @mankowitz give it a try to check if you are experiencing the same problem and it gives you a solution. In case it is, you can also mark this post as the solution. Thanks.

**Reply by eduardogr1984** [reply]
Yeah, just tried with the original post Channel and the example payload he attached and it worked, so it seems that the problem is maybe the same. 
Hope it helps someone else.

**Reply by ironcladmvtm** [reply]
@eduardogr1984 thank you for posting this solution. I have been fighting this for a few days now. Linking to this from my fourm.mirthproject.io [post](https://forums.mirthproject.io/forum/mirth-connect/support/181682-unable-to-receive-json-via-httplistener-source).

**Reply by eduardogr1984** [reply]
@ironcladmvtm  I'm glad it helped you. Was a big problem at that time for me and cost a lot of effort to guess the problem so I'm happy that this can help others :)

**Reply by ghust** [reply]
Four years later I want to reach out to say thank you for this post. It was driving me insane, and the trailing slash fixed it :D
